src/Utils/obiUtils.tsx

```diff
@@ -591,6 +597,114 @@ export async function createImportedActions(
+    }
+}
+
+/**
+ * Updates memory state for a single `TrainRound`.  Specifically, if we have an API action followed by
+ * a `SwitchCondition` node that consumes the output of that API, we know that each branch of the dialog
+ * should represent a scenario where the API has returned the given conditional value.  This code sets
+ * that conditional value in bot memory. 
+ */
+function setMemoryStateForApiActionWithSwitch(
+    entities: CLM.EntityBase[],
+    actions: CLM.ActionBase[],
+    scorerSteps: CLM.TrainScorerStep[],
+    scorerStepConditions: { [key: string]: CLM.Condition[] }) {
+    for (let i = 0; i < scorerSteps.length - 1; i = i + 1) {
+        if (!scorerSteps[i].labelAction) {
+            // Need a labelAction to figure out which action is associated with the scorer step.
+            continue
+        }
+        const j = i + 1
+        if (j >= scorerSteps.length) {
+            // There is no next action; we're done.
+            break
+        }
+        const actionId1 = scorerSteps[i].labelAction
+        const action1 = actions.find(a => a.actionId === actionId1)
+        if (!action1 || action1.actionType !== CLM.ActionTypes.API_LOCAL) {
+            // We only care about API actions followed by SwitchCondition-gated nodes.
+            continue
+        }
+        const scorerStep = scorerSteps[i]
+        const nextScorerStep = scorerSteps[j]
+        if (!nextScorerStep.importId || !scorerStepConditions[nextScorerStep.importId]) {
+            // The next action is not a SwitchCondition-gated node.
+            continue
+        }
+        // Get conditions from the SwitchCondition-gated node.
+        const conditions = scorerStepConditions[nextScorerStep.importId]
+        const enumConditionData = getEnumConditionData(entities, conditions)
+        // Set the logic result on the *current* scorer step.
+        if (!scorerStep.logicResult) {
+            scorerStep.logicResult = { logicValue: undefined, changedFilledEntities: [] }
+        }
+        for (const conditionEntity of enumConditionData) {
+            const filledEntity: CLM.FilledEntity = {
+                entityId: conditionEntity.enumEntityId,
+                values: [{
+                    userText: conditionEntity.enumValueText,
+                    displayText: conditionEntity.enumValueText,
+                    builtinType: null,
+                    resolution: null,
+                    enumValueId: conditionEntity.enumValueId
+                }]
+            }
+            scorerStep.logicResult.changedFilledEntities.push(filledEntity)
+        }
+    }
+}
+
+/**
+ * Gets the enum entity and enum value associated with the given conditions.
+ * @throws if the entity referenced in any `Condition` is not a valid enum.
+ */
+function getEnumConditionData(entities: CLM.EntityBase[], conditions: CLM.Condition[]): EnumDataFromCondition[] {
+    const output: EnumDataFromCondition[] = []
+    for (const condition of conditions) {
```

Comment:
It seems like this loop either throws error or adds `EnumDataFromCondition` which is the same as a `map`.

Seems like this could be reduced to function that extract enum data from a single condition and then expose it's mapping nature

`conditions.map(c => getEnumConditionData(c, entities))`

Reply:
JS patterns just not fully ingrained yet - updated